
<p align="center">
    <img src="https://github.com/johrollin/jupyter_variant_calling/blob/master/img/artcile_header.PNG" alt="Logo" width="600">
</p>

# About The Project

Recent developments in high-throughput sequencing (HTS) technologies and bioinformatics have drastically changed research on viral pathogens, especially for virus discovery and monitoring. Indeed, proper monitoring of the viral population requires information on the different isolates circulating in the studied area. For this purpose, HTS technologies have greatly facilitated the generation of new genomes of the detected viruses and their comparison. Nevertheless, the bioinformatics analyses allowing the reconstruction of genomes and the detection of Single Nucleotide Polymorphisms (SNPs) can potentially create bias, although it has not been widely addressed so far.  

Therefore, more knowledge is required on the limitation and possibility of predicting SNPs based on HTS-generated sequence datasets. To address this issue, we compared the ability of 14 plant virology laboratories, each employing a different bioinformatics pipeline, to detect 21 variants of pepino mosaic virus (PepMV) through large-scale Performance Testing (PT) using three artificially designed datasets. The bioinformatics analyses were divided into three key steps: reads pre-processing (quality trimming, merging …), virus identification (assembly, alignment, mapping …) and variant calling. Each step was evaluated independently through an original, step-by-step PT design with iteration between participants.   

Overall, this work underlines key parameters in SNP detection and proposes recommendations for reliable variant calling for plant viruses. The identification of the closest reference, mapping parameters and manual validation of the prediction were the most impactful analysis step for the success or failure of the predictions. Strategies to improve SNPs prediction are also discussed. 

For more information: https://www.biorxiv.org/content/XX 
<p align="center">
    <img src="https://github.com/johrollin/jupyter_variant_calling/blob/master/img/PT3_protocol_final2.png" alt="Logo" width="600">
</p>
